### PR TITLE
INTYGFV-14573: Logic to erase terminated customers data in Intygstjanst, webcert and intygsstatistik.

### PR DESCRIPTION
**Pull request by mww-sopra-steria**

### CODE REVIEW

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import se.inera.intyg.cts.domain.model.EraseService;
+import se.inera.intyg.cts.domain.model.Termination;
+import se.inera.intyg.cts.domain.model.TerminationStatus;
+import se.inera.intyg.cts.domain.repository.CertificateBatchRepository;
+import se.inera.intyg.cts.domain.repository.TerminationRepository;
+
+public class EraseDataForCareProvider {
```
Comment (IamFriberg) [inline customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java:15]: Missing @service annotation?

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  private final static Logger LOG = LoggerFactory.getLogger(EraseDataForCareProvider.class);
+
+  private final List<EraseDataInService> services;
+  private final CertificateBatchRepository certificateBatchRepository;
+  private final TerminationRepository terminationRepository;
+
+  public EraseDataForCareProvider(List<EraseDataInService> services,
```
Comment (IamFriberg) [inline customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java:23]: Some javadoc for these methods would be nice. Kind of hard to follow with all private methods.

File: customer-terminate-service/domain/src/test/java/se/inera/intyg/cts/domain/repository/InMemoryCertificateBatchRepository.java [hunk 3]
```diff
@@ -10,6 +11,12 @@ public class InMemoryCertificateBatchRepository implements CertificateBatchRepos
   private CertificateBatch certificateBatch;
   private List<CertificateText> certificateTexts;
 
+  @Override
+  public CertificateSummary certificateSummary(
+      Termination termination) {
```
Comment (IamFriberg) [inline customer-terminate-service/domain/src/test/java/se/inera/intyg/cts/domain/repository/InMemoryCertificateBatchRepository.java:16]: Is termination used in this method?

**Review by IamFriberg — CHANGES_REQUESTED** [review]
Som smaller changes. Could you update the description of the service in the readme?

File: core-stub-service/app/src/main/java/se/inera/intyg/css/application/api/IntygstjanstController.java [hunk 4]
```diff
@@ -39,4 +40,10 @@ public List<CertificateTextDTO> getCertificateTexts() {
     LOG.info("Get certificate texts.");
     return intygstjanstService.getCertificateTexts();
   }
+
+  @DeleteMapping("/certificates/{careProvider}")
+  public void eraseCareProvider(@PathVariable("careProvider") String careProviderId) {
+    LOG.info(String.format("Erase care provider '%s'.", careProviderId));
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply core-stub-service/app/src/main/java/se/inera/intyg/css/application/api/IntygstjanstController.java:46]: Wouldn't it be redundant, because the log message will come from core-stub-service which has stub in its name? :-)

File: core-stub-service/app/src/main/java/se/inera/intyg/css/application/api/StatistikController.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  private final static Logger LOG = LoggerFactory.getLogger(StatistikController.class);
+  private final StatistikService statistikService;
+
+  public StatistikController(StatistikService statistikService) {
+    this.statistikService = statistikService;
+  }
+
+  @DeleteMapping("/intygsidlist")
+  public void eraseCertificates(@RequestBody List<String> certificateIds) {
+    LOG.info(String.format("Erase certificateIds '%s'.", certificateIds));
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply core-stub-service/app/src/main/java/se/inera/intyg/css/application/api/StatistikController.java:25]: Wouldn't it be redundant, because the log message will come from core-stub-service which has stub in its name? :-)

File: core-stub-service/app/src/main/java/se/inera/intyg/css/application/api/WebcertController.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  private final static Logger LOG = LoggerFactory.getLogger(WebcertController.class);
+  private final WebcertService webcertService;
+
+  public WebcertController(WebcertService webcertService) {
+    this.webcertService = webcertService;
+  }
+
+  @DeleteMapping("/{careProvider}")
+  public void eraseCareProvider(@PathVariable("careProvider") String careProviderId) {
+    LOG.info(String.format("Erase care provider '%s'.", careProviderId));
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply core-stub-service/app/src/main/java/se/inera/intyg/css/application/api/WebcertController.java:24]: Wouldn't it be redundant, because the log message will come from core-stub-service which has stub in its name? :-)

File: core-stub-service/app/src/main/java/se/inera/intyg/css/testability/api/StatistikTestabilityController.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  private final StatistikTestabilityService statistikTestabilityService;
+
+  public StatistikTestabilityController(StatistikTestabilityService statistikTestabilityService) {
+    this.statistikTestabilityService = statistikTestabilityService;
+  }
+
+  @GetMapping("/erased/{certificateId}")
+  public boolean isCertificateErased(@PathVariable String certificateId) {
+    LOG.info(String.format("Check if certificate '%s' has been erased", certificateId));
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply core-stub-service/app/src/main/java/se/inera/intyg/css/testability/api/StatistikTestabilityController.java:27]: Wouldn't it be redundant, because the log message will come from core-stub-service which has stub in its name? :-)

File: core-stub-service/app/src/main/java/se/inera/intyg/css/testability/api/WebcertTestabilityController.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+  private final WebcertTestabilityService webcertTestabilityService;
+
+  public WebcertTestabilityController(WebcertTestabilityService webcertTestabilityService) {
+    this.webcertTestabilityService = webcertTestabilityService;
+  }
+
+  @GetMapping("/erased/{careProviderId}")
+  public boolean isCareProviderErased(@PathVariable String careProviderId) {
+    LOG.info(String.format("Check if care provider '%s' has been erased", careProviderId));
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply core-stub-service/app/src/main/java/se/inera/intyg/css/testability/api/WebcertTestabilityController.java:27]: Wouldn't it be redundant, because the log message will come from core-stub-service which has stub in its name? :-)

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/EraseController.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  
+  private final EraseService eraseService;
+
+  public EraseController(EraseService eraseService) {
+    this.eraseService = eraseService;
+  }
+
+  @PostMapping("")
+  void startErase() {
+    LOG.info("Start erase");
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/EraseController.java:24]: The endpoint is just a way to trigger 'Erase' instead of it being started by the EraseCareProviderTask. It will call the EraseService.erase() that will process/erase any termination that has the correct status.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java [hunk 10]
```diff
@@ -12,16 +12,20 @@
...
+import se.inera.intyg.cts.application.service.EraseServiceImpl;
 import se.inera.intyg.cts.application.service.TerminationService;
+import se.inera.intyg.cts.domain.model.TerminationId;
 
 @RestController
 @RequestMapping("/api/v1/terminations")
 public class TerminationController {
 
   private final TerminationService terminationService;
+  private final EraseServiceImpl eraseService;
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java:24]: Will fix!

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+import se.inera.intyg.cts.domain.model.TerminationId;
+import se.inera.intyg.cts.domain.service.EraseDataInService;
+import se.inera.intyg.cts.domain.service.EraseException;
+import se.inera.intyg.cts.infrastructure.persistence.entity.CertificateEntity;
+import se.inera.intyg.cts.infrastructure.persistence.entity.TerminationEntity;
+import se.inera.intyg.cts.infrastructure.persistence.repository.CertificateEntityRepository;
+import se.inera.intyg.cts.infrastructure.persistence.repository.TerminationEntityRepository;
+
+@Service
+public class EraseDataInIntygsstatistik implements EraseDataInService {
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java:32]: Will add javadoc for EraseDataInService!

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    this.scheme = scheme;
+    this.baseUrl = baseUrl;
+    this.port = port;
+    this.eraseCertificatesEndpoint = eraseCertificatesEndpoint;
+    this.eraseCareProviderEndpoint = eraseCareProviderEndpoint;
+    this.batchSize = batchSize;
+  }
+
+  @Override
+  public ServiceId serviceId() {
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java:69]: I'm not sure that it is possible to make it static when it is nterface method that has to implemented by its implementation class.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    return new ServiceId("intygsstatistik");
+  }
+
+  @Override
+  public void erase(Termination termination) throws EraseException {
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java:74]: There are dependencies between the calls. For example, all batches of certificates has to be successfully erased before the care provider is erased. Will give it a thought if I can refactor it a bit though.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/webcert/EraseDataInWebcert.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    } catch (Exception ex) {
+      LOG.error("Error calling webcert to delete care provider.", ex);
+      throw new EraseException(
+          String.format("Erase care provider failed with message '%s'", ex.getMessage())
+      );
+    }
+  }
+
+  @Override
+  public ServiceId serviceId() {
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/webcert/EraseDataInWebcert.java:63]: I'm not sure that it is possible to make it static when it is interface method that has to implemented by its implementation class.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/service/SendPackageNotification.java [hunk 15]
```diff
@@ -111,15 +114,15 @@ private boolean invalidEmailAddress(String emailAddress) {
...
 
   private void logSendEmailSuccess(String statusType, TerminationId terminationId) {
-    LOG.info("Successfully sent email {} for {}.", statusType, terminationId);
+    LOG.info("Successfully sent email {} for {}.", statusType, terminationId.id());
   }
 
-  private void logSendMessageFailure(String messageType, String statusType, TerminationId terminationId,
-      Exception e) {
-    LOG.error("Failure sending {} {} for {}.", messageType, statusType, terminationId, e);
+  private void logSendMessageFailure(String messageType, String statusType,
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/service/SendPackageNotification.java:124]: I just adjusted the terminationId in this method... The logic itself was not introduced in this PR. But I will see if we can do something.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/service/SendPasswordWithSMS.java [hunk 16]
```diff
@@ -10,40 +10,40 @@
...
+    try {
+      final var smsResponseDTO = sendSMS.sendSMS(formattedPhoneNumber, message);
+      logSendPasswordSuccess(termination, smsResponseDTO.job_id(), smsResponseDTO.log_href());
+      return true;
+
+    } catch (Exception e) {
+      logSendPasswordFailure(termination, e);
+      return false;
     }
+  }
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/service/SendPasswordWithSMS.java:40]: I just adjusted the terminationId in this method... The logic itself was not introduced in this PR. But I will see if we can do something.

File: customer-terminate-service/app/src/test/java/se/inera/intyg/cts/application/api/TerminationControllerTest.java [hunk 17]
```diff
@@ -25,77 +26,106 @@
...
-    @Mock
-    private TerminationService terminationService;
-    @InjectMocks
-    private TerminationController terminationController;
+  @Mock
+  private TerminationService terminationService;
 
-    private final TerminationDTO terminationDTO = TerminationTestDataBuilder.defaultTerminationDTO();
+  @Mock
+  private EraseServiceImpl eraseService;
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/test/java/se/inera/intyg/cts/application/api/TerminationControllerTest.java:41]: Will fix! I think there was a refactoring in IntelliJ that went wrong... :-)

File: customer-terminate-service/app/src/test/java/se/inera/intyg/cts/application/task/EraseCareProviderTaskTest.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import se.inera.intyg.cts.application.service.EraseServiceImpl;
+
+@ExtendWith(MockitoExtension.class)
+class EraseCareProviderTaskTest {
+
+  @Mock
+  private EraseServiceImpl eraseService;
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/test/java/se/inera/intyg/cts/application/task/EraseCareProviderTaskTest.java:17]: Will fix!

File: customer-terminate-service/app/src/test/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistikTest.java [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+        certificateEntityRepository, webClient, SCHEME, BASE_URL,
+        Integer.toString(mockIntygsstatistik.getPort()), ERASE_CERTIFICATE_ENDPOINT,
+        ERASE_CARE_PROVIDER_ENDPOINT, BATCH_SIZE);
+
+    termination = defaultTermination();
+    terminationEntity = toEntity(termination);
+  }
+
+  @AfterEach
+  void beforeEach() throws IOException {
```
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/test/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistikTest.java:86]: Will fix! :-)

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java (hunk 1 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java:15]: This is in the domain package and should not include any spring stuff. The bean is handled in the AppConfig.

Comment (IamFriberg) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java:15]: It seems like it is missing an interface if it is to be used as a service?

File: customer-terminate-service/app/src/test/java/se/inera/intyg/cts/application/task/EraseCareProviderTaskTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/test/java/se/inera/intyg/cts/application/task/EraseCareProviderTaskTest.java]: Fixed!

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java (hunk 1 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java]: Extracted interfaces for domain services that only had implementation classes.

File: customer-terminate-service/domain/src/test/java/se/inera/intyg/cts/domain/repository/InMemoryCertificateBatchRepository.java (hunk 3 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/test/java/se/inera/intyg/cts/domain/repository/InMemoryCertificateBatchRepository.java:16]: No, it is just the dummy implementation in the test package.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/application/api/TerminationController.java]: Fixed!

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java:70]: I've made the serviceId in the implementation classes constants. But as each class need to return its unique serviceId I can't make it static.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/webcert/EraseDataInWebcert.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/webcert/EraseDataInWebcert.java:64]: I've made the serviceId in the implementation classes constants. But as each class need to return its unique serviceId I can't make it static.

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java (hunk 2 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java]: Not sure that I agree... the private methods have meaningful names that say what they do. If we add javadoc to these methods it will be additional maintenance overhead as the javadoc always have to be consistent with the code and any internal refactoring of the class will break it.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java:32]: Javadoc has been added to EraseDataInService.

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/service/SendPasswordWithSMS.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/service/SendPasswordWithSMS.java:40]: I've looked at the implementation and I think they add readability to the code. I will leave it as is.

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java (hunk 2 above)
Comment (IamFriberg) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java]: At least for me it would help to get a quick summary of the purpose each of these methods. Javadoc is best practise and used in all public libraries to help developers get a understanding of what the methods do. The overhead of updating is minimal and should be a part of the proccess. We can not asume that the purpose of the code is clear and future developers can quickly understand what the intended purpose is. One of my most common issues with working with older code is that the developers before me has made no dokumentation whatsoever and that creates a lot of extra work. It takes just a small amount of time to create a summary when you know and develop the code compared to trying to understand it 5 years later. :) The meaning of refactoring is to change the strukture of the code and not the behaviour so it should not affect the Javadoc and if it does it just takes a small amount of effort to update it. :)

**Review by IamFriberg — CHANGES_REQUESTED** [review]
Se comment :)

File: customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mww-sopra-steria) [reply customer-terminate-service/app/src/main/java/se/inera/intyg/cts/infrastructure/integration/intygsstatistik/EraseDataInIntygsstatistik.java:32]: I've added some specific javadoc for the implementation in Intygsstatistik, as it is a bit more complex than the other services.

File: customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java (hunk 2 above)
Comment (mww-sopra-steria) [reply customer-terminate-service/domain/src/main/java/se/inera/intyg/cts/domain/service/EraseDataForCareProvider.java]: OK, I understand and get your point. I was mistaken before that you were asking for javadoc on the private methods.

I've written javadoc and tried to describe the behaviour of the method without going in to too much details.

**Review by IamFriberg — APPROVED** [review]
Snyggt!
